control/man/imageinfo.1
.TH IMAGEINFO "1" "January 2014" "AstropPhotography tools" "User Commands"
.SH NAME
imageinfo \- display information about an image
.SH SYNOPSIS
.B imageinfo [ options ]
.I images
...
.I 
.SH DESCRIPTION
The 
.BR imageinfo (1)
command displays information about pixel values of each image file
specified on the command line.
It displays minimum, mean, median
and maximum value. In addition, the number of NaNs is displayed
if the pixel type allows for NaNs.
.SH OPTIONS
.TP
.B \-c,\-\-config= config
Use a configuration detabase different from the default
.IR ~/.astrophoto.db .
.TP
.B \-d,\-\-debug
increase the debug level.
.TP
.B \-f,\-\-fits
In addition to the pixel related information, also show the information
contained in the FITS headers.
.TP
.B \-h,\-\-help
Display a help message about the available commands and exit.
.TP
.BI \-r,\-\-repo= repository
Instead of from files, read image information from the image repository
named
.IR repository .
In this case, the 
.I images
arguments are ids (numbers) specifying the images in the repository.
There are no glob pattern matching available for these ids, instead
one should use tools like 
.BR seq (1)
to produce sequences of ids.
.TP
.B \-v,\-\-verbose
Verbose display. In addition to the pixel value information, information
about the pixel type and image depth is displayed in verbose display.
For the meaning of the attributes see the section ATTRIBUTES below.
.SH ATTRIBUTES
The results of imageinfo are displayed as a list of attribute=value
pairs. The following attributes are available, some only in verbose
mode. Information only displayed in verbose mode is marked by a (V).

.TP
Bpp
The number of bytes per pixel (V).
.TP
bpp
The number of bits per pixel (V).
.TP
max
The maximum of the actual values of the image.
.TP
maxvalue
The maximum possible value of a pixel (V).
.TP
mean
The mean of the actual pixel values of the image.
.TP
median
The median of the actual pixel values of the image.
.TP
min
The minimum of the actual values of the image.
.TP
minvalue
The minimum possible value of a pixel (V).
.TP
nans
Number of nans of a floating point image. They usually correspond
to bad pixels. When images are transformed, some areas may not be covered
by image data and a are filled in with Nans. In these cases, an excessive
number of nans is displayed. Using Nans ensures that these pixels are
never used to compute any displayable pixel.
This is attribute is not displayed for images with integer pixels.
.TP
pixel_type
Type of pixel. This is the C++ type identifier of the pixel type (V).
.TP
planes
The number of planes, typically 1 for raw and monochrome images, 
3 for processed RGB images.
.SH FILES
.IR ~/.astrophoto.db
.SH "SEE ALSO"
.BR astrod (1)
.SH AUTHORS
Prof. Dr. Andreas Mueller


control/tools/imageinfo.cpp
/*
 * imageinfo.cpp -- display information about images
 *
 * (c) 2013 Prof Dr Andreas Mueller, Hochschule Rapperswil
 */
#include <includes.h>
#include <AstroImage.h>
#include <AstroIO.h>
#include <AstroFilter.h>
#include <AstroFilterfunc.h>
#include <AstroDebug.h>
#include <stdexcept>
#include <AstroUtils.h>
#include <AstroConfig.h>
#include <AstroProject.h>

using namespace astro;
using namespace astro::image;
using namespace astro::image::filter;
using namespace astro::project;
using namespace astro::config;
using namespace astro::io;

namespace astro {
namespace app {
namespace imageinfo {

static std::string	reponame;

bool	verbose = false;
bool	fitsinfo = false;

/**
 * \brief Common image info display function
 */
int	show_imageinfo(ImagePtr image) {
	if (fitsinfo) {
		std::cout << *image;
	}

	if (verbose) {
		std::cout << "bpp=" << image->bitsPerPixel();
		std::cout << " Bpp=" << image->bytesPerPixel();
		std::cout << " planes=" << image->planes();
		std::cout << " minvalue=" << image->minimum();
		std::cout << " maxvalue=" << image->maximum();
		std::cout << " pixel_type=";
		std::cout << demangle(image->pixel_type().name());
		std::cout << " ";
	}

	// find maximum, minimum, average and median values
	double	maximum = astro::image::filter::max(image);
	double	minimum = astro::image::filter::min(image);
	double	mean = astro::image::filter::mean(image);
	double	median = astro::image::filter::median(image);
	double	nans = -1;
	try {
		nans = astro::image::filter::countnans(image);
	} catch (...) {
		debug(LOG_DEBUG, DEBUG_LOG, 0, "ignoring nans");
	}
	
	std::cout << "min=" << minimum;
	if (median < mean) {
		std::cout << " median=" << median;
		std::cout << " mean=" << mean;
	} else {
		std::cout << " mean=" << mean;
		std::cout << " median=" << median;
	}
	std::cout << " max=" << maximum;
	if (nans >= 0) {
		std::cout << " nans=" << nans;
	}

	return EXIT_SUCCESS;
}

/**
 * \brief Show image info for an image file
 */
int	show_imagefile(const std::string& filename) {
	debug(LOG_DEBUG, DEBUG_LOG, 0, "image info for: %s", filename.c_str());
	try {
		std::cout << filename << ": ";
		FITSin	infile(filename);
		ImagePtr	image = infile.read();

		return show_imageinfo(image);
	} catch (std::exception& x) {
		std::cerr << "could not process " << filename << ": ";
		std::cerr << x.what() << std::endl;
	}
	return EXIT_FAILURE;
}

/**
 * \brief Display information about an image found in the repository
 */
int	show_imagerepo(const std::string& argument) {
	try {
		ImageRepoPtr	repo = Configuration::get()->repo(reponame);
		long	imageid = std::stol(argument);
		std::cout << imageid << ": ";
		ImagePtr	image = repo->getImage(imageid);
		return show_imageinfo(image);
	} catch (std::exception& x) {
		std::cerr << "could not process " << argument;
		std::cerr << " in repo " << reponame << ": " << x.what();
		std::cerr << std::endl;
	}
	return EXIT_FAILURE;
}

/**
 * \brief usage function for the imageinfo program
 */
void	usage(const std::string& progname) {
	std::cout << "Usage:" << std::endl;
	Path	path(progname);
	std::cout << path.basename() << " [ options ] images ..." << std::endl;
	std::cout << "Display information about pixel values of FITS images "
		"specified by their" << std::endl;
	std::cout << "file names in the images arguments." << std::endl;
	std::cout << "Options:" << std::endl;
	std::cout << "  -c,--config=cfg    use configuration database <cfg> "
		"instead of" << std::endl;
	std::cout << "                     default ~/astrophoto.db";
	std::cout << std::endl;
	std::cout << "  -d,--debug         increase debug level" << std::endl;
	std::cout << "  -f,--fits          display FITS header info" << std::endl;
	std::cout << "  -h,--help          display help message end exit";
	std::cout << std::endl;
	std::cout << "  -r,--repo=<repo>   use image repository named <repo> instead of direct" << std::endl;
	std::cout << "                     file access. In this case, the images arguments are" << std::endl;
	std::cout << "                     numbers (ids) of images in the repository, not file" << std::endl;
	std::cout << "                     names" << std::endl;;
	std::cout << "  -v,--verbose       verbose mode, show additional "
		"information about" << std::endl;
	std::cout << "                     pixel types" << std::endl;
}

/**
 * \brief Table of options
 */
static struct option    longopts[] = {
/* name         argument?               int*            int */
{ "config",	required_argument,	NULL,		'c' }, /* 0 */
{ "debug",	no_argument,		NULL,		'd' }, /* 1 */
{ "help",	no_argument,		NULL,		'h' }, /* 2 */
{ "repo",	required_argument,	NULL,		'r' }, /* 3 */
{ "verbose",	no_argument,		NULL,		'v' }, /* 4 */
{ "fits",	no_argument,		NULL,		'f' }, /* 5 */
{ NULL,         0,			NULL,		0   }
};

/**
 * \brief main function
 */
int	main(int argc, char *argv[]) {
	int	c;
	int	longindex;
	while (EOF != (c = getopt_long(argc, argv, "dr:", longopts,
		&longindex)))
		switch (c) {
		case 'c':
			Configuration::set_default(optarg);
			break;
		case 'd':
			debuglevel = LOG_DEBUG;
			break;
		case 'f':
			fitsinfo = true;
			break;
		case 'h':
			usage(std::string(argv[0]));
			return EXIT_FAILURE;
		case 'r':
			reponame = optarg;
			break;
		case 'v':
			verbose = true;
			break;
		}

	if (argc <= optind) {
		debug(LOG_ERR, DEBUG_LOG, 0, "no image file arguments");
		throw std::runtime_error("not image file arguments");
	}

	for (; optind < argc; optind++) {
		std::string	argument(argv[optind]);
		if (reponame.size() != 0) {
			show_imagerepo(argument);
		} else {
			show_imagefile(argument);
		}
		std::cout << std::endl;
	}
	
	return EXIT_SUCCESS;
}

} // namespace imageinfo
} // namespace app
} // namespace astro

int	main(int argc, char *argv[]) {
	return astro::main_function<astro::app::imageinfo::main>(argc, argv);
}
